messagix/client.go

```diff
@@ -388,3 +393,28 @@ func (c *Client) GetTaskId() int {
 	c.activeTasks = append(c.activeTasks, id)
 	return id
 }
+
+func (c *Client) EnableSendingMessages() {
+	c.sendMessagesCond.L.Lock()
+	c.canSendMessages = true
+	c.sendMessagesCond.Broadcast()
+	c.sendMessagesCond.L.Unlock()
+}
+
+func (c *Client) WaitUntilCanSendMessages(timeout time.Duration) error {
+	c.sendMessagesCond.L.Lock()
+	defer c.sendMessagesCond.L.Unlock()
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for !c.canSendMessages {
+		select {
+		case <-timer.C:
+			return fmt.Errorf("timeout waiting for canSendMessages")
+		default:
+			c.sendMessagesCond.Wait()
+		}
+	}
+	return nil
+}
```

Comment:
I think this would solve the two issues mentioned above (the blocking forever and the deadlock), wdyt @tulir?
```suggestion
func (c *Client) WaitUntilCanSendMessages(timeout time.Duration) error {
	timer := time.NewTimer(timeout)
	defer timer.Stop()

	done := make(chan struct{})
	go func() {
		c.sendMessagesCond.L.Lock()
		defer c.sendMessagesCond.L.Unlock()
		c.sendMessagesCond.Wait()
		close(done)
	}()

	for !c.canSendMessages {
		select {
		case <-timer.C:
			return fmt.Errorf("timeout waiting for canSendMessages")
		case <-done:
			if c.canSendMessages {
				return nil
			}
		}
	}
	return nil
}
```

Reply:
Hmm, I think it still has the deadlock issue: the lock is acquired during `Wait`, but the function doing `Broadcast` also wants the lock

Also, if the flag was theoretically set to false in between Wait returning and the `for` loop checking the condition, then it would get in a tight loop until the timer expires (because `done` would be closed, but the `canSendMessages` check would fail and not return). Right now that never happens because the flag is never set to false, but it should probably be set to false when disconnecting 🤔

This is getting so complicated that I'm thinking it might be better to just have retries with dumb exponential backoff in between (like max 5 attempts and sleep 2+4+8+16 seconds between attempts)

Reply:
> Hmm, I think it still has the deadlock issue: the lock is acquired during Wait, but the function doing Broadcast also wants the lock

From https://github.com/golang/go/blob/master/src/sync/cond.go#L66:

> Wait atomically unlocks c.L and suspends execution of the calling goroutine. After later resuming execution, Wait locks c.L before returning. Unlike in other systems, Wait cannot return unless awoken by Broadcast or Signal.

I think this prevent the deadlock you were describing, doesn't it?

> Right now that never happens because the flag is never set to false, but it should probably be set to false when disconnecting 🤔

Why should we do that? Client is always [disposed on User.unlockedDisconnect](https://github.com/mautrix/meta/blob/main/user.go#L1196).

Reply:
> I think this prevent the deadlock you were describing, doesn't it?

Oh yes, so that part is fine

> Why should we do that? Client is always [disposed on User.unlockedDisconnect](https://github.com/mautrix/meta/blob/main/user.go#L1196).

Same as why #64 needs to cancel requests when disconnecting: `messagix.Client` handles normal network errors internally and reconnects automatically, such errors don't call `User.Disconnect`

Reply:
That makes sense!

Just added a `disableSendingMessages` that gets called every time we reconnect the socket:

https://github.com/mautrix/meta/pull/63/commits/c93d8930eb4079e5495a733f2e02a8b5b395e45c